include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -164,12 +226,58 @@ class NBL_API IDescriptorSetLayout : public virtual core::IReferenceCounted
 		IDescriptorSetLayout(const SBinding* const _begin, const SBinding* const _end) : 
 			m_bindings((_end-_begin) ? core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<SBinding>>(_end-_begin) : nullptr)
 		{
+			struct SBindingRedirectBuildInfo
+			{
+				uint32_t binding;
+				uint32_t count;
+
+				inline bool operator< (const SBindingRedirectBuildInfo& other) const { return binding < other.binding; }
+			};
+
+			core::vector<SBindingRedirectBuildInfo> buildInfo_descriptors[asset::EDT_COUNT];
+			core::vector<SBindingRedirectBuildInfo> buildInfo_samplers;
+
+			for (auto b = _begin; b != _end; ++b)
+			{
+				buildInfo_descriptors[b->type].emplace_back(b->binding, b->count);
+				if (b->type == EDT_COMBINED_IMAGE_SAMPLER && b->samplers == nullptr)
+				{
+					buildInfo_samplers.emplace_back(b->binding, b->count);
+					m_mutableSamplerCount += b->count;
+				}
+			}
+
+			for (auto type = 0u; type < asset::EDT_COUNT; ++type)
+				m_descriptorRedirects[type] = SBindingRedirect(buildInfo_descriptors[type].size());
+			m_samplerRedirects = SBindingRedirect(buildInfo_samplers.size());
+
+			auto buildRedirect = [](SBindingRedirect& redirect, core::vector<SBindingRedirectBuildInfo>& info)
+			{
+				std::sort(info.begin(), info.end());
+
+				for (size_t i = 0u; i < info.size(); ++i)
+				{
+					redirect.bindings[i] = info[i].binding;
+					redirect.offsets[i] = info[i].count;
+				}
+
+				std::exclusive_scan(redirect.offsets, redirect.offsets + info.size(), redirect.offsets, 0u);
+			};
+
+			for (auto type = 0u; type < asset::EDT_COUNT; ++type)
+				buildRedirect(m_descriptorRedirects[type], buildInfo_descriptors[type]);
+			buildRedirect(m_samplerRedirects, buildInfo_samplers);
```

Comment:
this should really be folded into the constructor of `SBindingRedirect`

Reply:
`m_mutableSamplerCount` you can get from the inclusive scan later (lets switch away from exclusive because the first element is always 0 there)